api/docs/multi_trace_window.dox

```diff
@@ -0,0 +1,396 @@
+  init and exit.
+
++ Each trace buffer has a corresponding externally_written variable holding a
+  distance into the buffer that was written out by another thread.
+
++ On hitting the trace window endpoint threshold, the triggering thread grabs the
+  mutex and walks the buffers.
+
+  The triggerer doesn't have the current buffer position pointer.  Instead it walks
+  the buffer until it reaches zeroed memory (we zero the buffer after each flush).
+  We have no synchronization with the owning thread: but observing writes out of
+  order should be ok since we'll just miss one by stopping too early.  We need to fix
+  things up in post-processing in any case, because we need the phase transition to
+  be at a clean point (we can't identify that point from triggerer: if we end at an
+  instr entry, we don't know if some memrefs are coming afterward or not).  In
+  post-processing we adjust that position to the end of the block, and we split the
+  buffer contents around that point to the neighboring traces.
+
+  The triggerer does a store-release of the furthest-writting point into the
+  externally_written variable.
+
++ Whenever a trace writes out its buffer, it does a load-acquire on the
+  externally_written variable and if non-zero it writes out a marker in the buffer
+  header.  Post-processing reads the marker and uses it to split the buffer at the
+  nearest block boundary after the marker value.
+
++ If windows are small enough that the triggerer doesn't complete its buffer walk
+  before a new window starts: other thread buffers may completely go into the new
+  window.  That seems ok: if the windows are that small, in the absence of
+  application synchronization the resulting window split should be a possible thread
+  ordering.
+
+This scheme ends up with block-level accuracy since the trigger thread's marked
+transition point must be adjusted to essentially a block boundary in post-processing.
+Thus, it does not seem any better than the other schemes, and it is more complex.
+
+# Online Traces
+
+It makes sense for offline to treat each window trace as separate and simulate them
+separately (though aggregating the results to cover the whole application).
+
+But for online: we have the same instance of the simulator or analysis tool
+throughout the whole application run.  It will get confused if it throws away thread
+bookkeeping on a thread exit for a window.
+
+Either we have a window-controller simulator who spins up and down a new instance of
+the real target simulator/tool on each window, or we introduce new "end phase/start
+phase" markers.  If we have split offline traces, those would only be for online
+though which does not sound appealing.  Simulators/tools would need special handling
+for them: reporting statistics for the phase while continuing to aggregate for a
+multi-phase report or something.
+
+We might want combined files for offline too, as discussed above.  That would unify
+the two, which is appealing.
+
+
```

Comment:
Maybe add a short conclusion section that summarizes the design taken.

Reply:
Done.